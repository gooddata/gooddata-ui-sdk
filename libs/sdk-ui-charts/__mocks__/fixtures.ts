// (C) 2007-2019 GoodData Corporation
import { LegacyExecutionRecording, legacyRecordedDataView } from "@gooddata/sdk-backend-mockingbird";

import * as rec from "./recordings/playlist.js";
import { DataViewFacade } from "@gooddata/sdk-ui";

function legacyRecordedDataFacade(rec: LegacyExecutionRecording): DataViewFacade {
    return DataViewFacade.for(legacyRecordedDataView(rec));
}

//
// new fixtures
//

export const testWorkspace = "testWorkspace";

//
// Area chart fixtures
//

export const areaChartWith3MetricsAndViewByAttribute = legacyRecordedDataFacade(
    rec.AreaChartWith3MetricsAndViewByAttribute,
);

export const areaChartWithMeasureViewByAndStackBy = legacyRecordedDataFacade(
    rec.AreaChartWithMeasureViewByAndStackBy,
);

//
// Bar chart fixtures
//

export const barChartWith4MetricsAndViewByTwoAttributes = legacyRecordedDataFacade(
    rec.BarChartWith4MetricsAndViewByTwoAttributes,
);
export const barChartWithStackByAndViewByAttributes = legacyRecordedDataFacade(
    rec.BarChartWithStackByAndViewByAttributes,
);
export const barChartWithViewByAttribute = legacyRecordedDataFacade(rec.BarChartWithViewByAttribute);

export const barChartWith3MetricsAndViewByAttribute = legacyRecordedDataFacade(
    rec.BarChartWith3MetricsAndViewByAttribute,
);
export const barChartWith3MetricsAndViewByAttributeFunformat = legacyRecordedDataFacade(
    rec.BarChartWith3MetricsAndViewByAttributeFunformat,
);

export const barChartWith3MetricsAndViewByAttributePercInFormat = legacyRecordedDataFacade(
    rec.BarChartWith3MetricsAndViewByAttributePercInFormat,
);

export const barChartWithPopMeasureAndViewByAttribute = legacyRecordedDataFacade(
    rec.BarChartWithPopMeasureAndViewByAttribute,
);

export const barChartWithPopMeasureAndViewByAttributeX6 = legacyRecordedDataFacade(
    rec.BarChartWithPopMeasureAndViewByAttributeX6,
);

export const barChartWithPreviousPeriodMeasure = legacyRecordedDataFacade(
    rec.BarChartWithPreviousPeriodMeasure,
);
export const barChartWithSingleMeasureAndNoAttributes = legacyRecordedDataFacade(
    rec.BarChartWithSingleMeasureAndNoAttributes,
);

export const barChartWithPreviousPeriodMeasureX6 = legacyRecordedDataFacade(
    rec.BarChartWithPreviousPeriodMeasureX6,
);

//
// Bubble chart
//
export const bubbleChartWith2MetricsAndAttributeNoPrimaries = legacyRecordedDataFacade(
    rec.BubbleChartWith2MetricsAndAttributeNoPrimaries,
);

export const bubbleChartWith1Metric = legacyRecordedDataFacade(rec.BubbleChartWith1Metric);

export const bubbleChartWith3MetricsAndAttribute = legacyRecordedDataFacade(
    rec.BubbleChartWith3MetricsAndAttribute,
);
export const bubbleChartWith3MetricsAndAttributeNullsInData = legacyRecordedDataFacade(
    rec.BubbleChartWith3MetricsAndAttributeNullsInData,
);

//
// Combo chart
//
export const comboWithTwoMeasuresAndViewByAttribute = legacyRecordedDataFacade(
    rec.ComboChartWithTwoMeasuresViewByAttribute,
);

export const comboChartWithTwoMeasuresViewByAttributeNoBuckets = legacyRecordedDataFacade(
    rec.ComboChartWithTwoMeasuresViewByAttributeNoBuckets,
);

export const comboChartWithTwoMeasuresViewByAttributePercformat = legacyRecordedDataFacade(
    rec.ComboChartWithTwoMeasuresViewByAttributePercformat,
);

//
//
//
export const headlineWithOneMeasure = legacyRecordedDataFacade(rec.HeadlineWithOneMeasure);
export const headlineWithOneMeasureWithIdentifier = legacyRecordedDataFacade(
    rec.HeadlineWithOneMeasureWithIdentifier,
);
export const headlineWithTwoMeasures = legacyRecordedDataFacade(rec.HeadlineWithTwoMeasures);
export const headlineWithTwoMeasuresWithIdentifier = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresWithIdentifier,
);
export const headlineWithTwoMeasuresFirstEmpty = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresFirstEmpty,
);
export const headlineWithTwoMeasuresSecondEmpty = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresSecondEmpty,
);
export const headlineWithTwoMeasuresBothEmpty = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresBothEmpty,
);
export const headlineWithTwoMeasuresBothZero = legacyRecordedDataFacade(rec.HeadlineWithTwoMeasuresBothZero);
export const headlineWithTwoMeasuresFirstZero = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresFirstZero,
);
export const headlineWithTwoMeasuresSecondZero = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresSecondZero,
);
export const headlineWithTwoMeasuresBothSame = legacyRecordedDataFacade(rec.HeadlineWithTwoMeasuresBothSame);

//
// Heatmap recordings
//

export const heatMapWithEmptyCells = legacyRecordedDataFacade(rec.HeatMapWithEmptyCells);
export const heatMapWithMetricRowColumn = legacyRecordedDataFacade(rec.HeatMapWithMetricRowColumn);
//
// Pie chart recordings
//

export const pieChartWithMetricsOnly = legacyRecordedDataFacade(rec.PieChartWithMetricsOnly);
export const pieChartWithMetricsOnlyFundata = legacyRecordedDataFacade(rec.PieChartWithMetricsOnlyFundata);

//
// Scatter plot recording
//

export const scatterPlotWith2MetricsAndAttributeNullsInData = legacyRecordedDataFacade(
    rec.ScatterPlotWith2MetricsAndAttributeNullsInData,
);
export const scatterPlotWith2MetricsAndAttributeWithPrimary = legacyRecordedDataFacade(
    rec.ScatterPlotWith2MetricsAndAttributeWithPrimary,
);

//
// Treemap recordings
//

export const treemapWithMetricAndStackByAttribute = legacyRecordedDataFacade(
    rec.TreemapWithMetricAndStackByAttribute,
);
export const treemapWithMetricAndViewByAttribute = legacyRecordedDataFacade(
    rec.TreemapWithMetricAndViewByAttribute,
);
export const treemapWithTwoMetricsAndStackByAttribute = legacyRecordedDataFacade(
    rec.TreemapWithTwoMetricsAndStackByAttribute,
);
export const treemapWithMetricViewByAndStackByAttribute = legacyRecordedDataFacade(
    rec.TreemapWithMetricViewByAndStackByAttribute,
);
