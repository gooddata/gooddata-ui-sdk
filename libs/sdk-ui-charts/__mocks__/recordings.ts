// (C) 2020 GoodData Corporation

import { DataViewFacade } from "@gooddata/sdk-ui";
import { ScenarioRecording, recordedDataView } from "@gooddata/sdk-backend-mockingbird";

export function recordedDataFacade(rec: ScenarioRecording): DataViewFacade {
    return DataViewFacade.for(recordedDataView(rec));
}
