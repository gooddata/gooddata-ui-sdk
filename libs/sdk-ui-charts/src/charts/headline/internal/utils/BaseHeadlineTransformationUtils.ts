// (C) 2023 GoodData Corporation
import { IDataView } from "@gooddata/sdk-backend-spi";
import {
    convertDrillableItemsToPredicates,
    DataViewFacade,
    ExplicitDrill,
    HeadlineElementType,
    isSomeHeaderPredicateMatched,
} from "@gooddata/sdk-ui";
import { IMeasureDescriptor } from "@gooddata/sdk-model";

import BaseHeadlineDataItem from "../headlines/baseHeadline/baseHeadlineDataItems/BaseHeadlineDataItem.js";
import {
    createHeadlineDataItem,
    getExecutionData,
    IHeadlineExecutionData,
} from "./HeadlineTransformationUtils.js";
import { IBaseHeadlineData, IBaseHeadlineItem } from "../interfaces/BaseHeadlines.js";
import { IHeadlineDataItem } from "../interfaces/Headlines.js";

export function getBaseHeadlineData(dataView: IDataView, drillableItems: ExplicitDrill[]): IBaseHeadlineData {
    const drillablePredicates = convertDrillableItemsToPredicates(drillableItems);

    const dv = DataViewFacade.for(dataView);
    const executionData = getExecutionData(dv);
    const itemHeaders = dv.meta().measureDescriptors();

    const primaryItem = createBaseHeadlineItem(
        executionData[0],
        isSomeHeaderPredicateMatched(drillablePredicates, itemHeaders[0], dv),
        "primaryValue",
    );

    let secondaryItemData: IHeadlineExecutionData;
    let secondaryItemHeader: IMeasureDescriptor;
    let tertiaryItemData: IHeadlineExecutionData;
    let tertiaryItemHeader: IMeasureDescriptor;
    if (executionData.length === 3) {
        // There are 2 secondary metrics
        // The left item will be the second metric
        // The right item will be the third metric
        secondaryItemData = executionData[2];
        secondaryItemHeader = itemHeaders[2];
        tertiaryItemData = executionData[1];
        tertiaryItemHeader = itemHeaders[1];
    } else {
        secondaryItemData = executionData[1];
        secondaryItemHeader = itemHeaders[1];
    }

    const secondaryItem = createBaseHeadlineItem(
        secondaryItemData,
        secondaryItemHeader && isSomeHeaderPredicateMatched(drillablePredicates, secondaryItemHeader, dv),
        "secondaryValue",
    );

    const tertiaryItem = createBaseHeadlineItem(
        tertiaryItemData,
        tertiaryItemHeader && isSomeHeaderPredicateMatched(drillablePredicates, tertiaryItemHeader, dv),
        "secondaryValue",
    );

    return {
        primaryItem,
        secondaryItem,
        tertiaryItem,
    };
}

export function createBaseHeadlineItem(
    executionData: IHeadlineExecutionData,
    isDrillable: boolean,
    elementType: HeadlineElementType,
): IBaseHeadlineItem<IHeadlineDataItem> {
    const data = createHeadlineDataItem(executionData, isDrillable);

    return data
        ? {
              data,
              elementType,
              baseHeadlineDataItemComponent: BaseHeadlineDataItem,
          }
        : null;
}
