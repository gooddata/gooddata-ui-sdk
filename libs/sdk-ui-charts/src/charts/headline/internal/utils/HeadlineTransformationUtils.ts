// (C) 2007-2019 GoodData Corporation
import cloneDeep = require("lodash/cloneDeep");
import get = require("lodash/get");
import isEmpty = require("lodash/isEmpty");
import isNumber = require("lodash/isNumber");
import { DataValue, DataViewFacade, IDataView, IMeasureDescriptor } from "@gooddata/sdk-backend-spi";
import * as invariant from "invariant";
import { IntlShape } from "react-intl";
import {
    IDrillEvent,
    IDrillEventContextHeadline,
    IDrillEventIntersectionElement,
    IHeaderPredicate,
    isSomeHeaderPredicateMatched,
    HeadlineElementType,
    VisualizationTypes,
} from "@gooddata/sdk-ui";
import { IHeadlineData, IHeadlineDataItem } from "../../Headlines";
import { Identifier } from "@gooddata/sdk-model";

export interface IHeadlineExecutionData {
    measureHeaderItem: IMeasureDescriptor["measureHeaderItem"];
    value: DataValue;
}

export interface IHeadlineDrillItemContext {
    localIdentifier: Identifier;
    value: string;
    element: HeadlineElementType;
}

function createHeadlineDataItem(executionDataItem: IHeadlineExecutionData): IHeadlineDataItem {
    if (!executionDataItem) {
        return null;
    }

    return {
        localIdentifier: executionDataItem.measureHeaderItem.localIdentifier,
        title: executionDataItem.measureHeaderItem.name,
        value: executionDataItem.value ? String(executionDataItem.value) : null,
        format: executionDataItem.measureHeaderItem.format,
        isDrillable: false,
    };
}

function createTertiaryItem(executionData: IHeadlineExecutionData[], intl: IntlShape): IHeadlineDataItem {
    const secondaryHeaderItem = get(executionData, [1, "measureHeaderItem"]);
    if (!secondaryHeaderItem) {
        return null;
    }

    const primaryValueString = get(executionData, [0, "value"]);
    const primaryValue = primaryValueString !== null ? Number(primaryValueString) : null;
    const secondaryValueString = get(executionData, [1, "value"]);
    const secondaryValue = secondaryValueString !== null ? Number(secondaryValueString) : null;

    const tertiaryTitle = intl.formatMessage({ id: "visualizations.headline.tertiary.title" });

    const isCountableValue = isNumber(primaryValue) && isNumber(secondaryValue);
    const tertiaryValue =
        isCountableValue && secondaryValue !== 0
            ? ((primaryValue - secondaryValue) / secondaryValue) * 100
            : null;

    return {
        localIdentifier: "tertiaryIdentifier",
        title: tertiaryTitle,
        value: tertiaryValue !== null ? String(tertiaryValue) : null,
        format: null,
        isDrillable: false,
    };
}

function getExecutionData(dv: DataViewFacade): IHeadlineExecutionData[] {
    const headerItems = dv.measureDescriptors();
    const data = dv.singleDimData();

    return headerItems.map((item, index) => {
        const value = data[index];

        invariant(value !== undefined, "Undefined execution value data for headline transformation");
        invariant(item.measureHeaderItem, "Missing expected measureHeaderItem");

        return {
            measureHeaderItem: item.measureHeaderItem,
            value,
        };
    });
}

/**
 * Get {HeadlineData} used by the {Headline} component.
 *
 * @param dataView - data to visualize
 * @param intl - Required localization for compare item title
 * @returns {*}
 */
export function getHeadlineData(dataView: IDataView, intl: IntlShape): IHeadlineData {
    const dv = new DataViewFacade(dataView);
    const executionData = getExecutionData(dv);

    const primaryItem = createHeadlineDataItem(executionData[0]);

    const secondaryItem = createHeadlineDataItem(executionData[1]);
    const secondaryItemProp = secondaryItem ? { secondaryItem } : {};

    const tertiaryItem = createTertiaryItem(executionData, intl);
    const tertiaryItemProp = tertiaryItem ? { tertiaryItem } : {};

    return {
        primaryItem,
        ...secondaryItemProp,
        ...tertiaryItemProp,
    };
}

/**
 * Take headline data and apply list of drillable items.
 * The method will return copied collection of the headline data with altered drillable status.
 *
 * @param headlineData - The headline data that we want to change the drillable status.
 * @param drillableItems - list of drillable items {uri, identifier}
 * @param dataView - data visualized by the headline
 * @returns altered headlineData
 */
export function applyDrillableItems(
    headlineData: IHeadlineData,
    drillableItems: IHeaderPredicate[],
    dataView: IDataView,
): IHeadlineData {
    const dv = new DataViewFacade(dataView);
    const data = cloneDeep(headlineData);
    const { primaryItem, secondaryItem } = data;
    const [primaryItemHeader, secondaryItemHeader] = dv.measureDescriptors();

    if (!isEmpty(primaryItem) && !isEmpty(primaryItemHeader)) {
        primaryItem.isDrillable = isSomeHeaderPredicateMatched(drillableItems, primaryItemHeader, dv);
    }

    if (!isEmpty(secondaryItem) && !isEmpty(secondaryItemHeader)) {
        secondaryItem.isDrillable = isSomeHeaderPredicateMatched(drillableItems, secondaryItemHeader, dv);
    }

    return data;
}

/**
 * Build drill event data (object with execution and drill context) from the data obtained by clicking on the {Headline}
 * component an from the execution objects.
 *
 * @param itemContext - data received from the click on the {Headline} component.
 * @param dataView - data visualized by the headline
 * @returns {*}
 */
export function buildDrillEventData(
    itemContext: IHeadlineDrillItemContext,
    dataView: IDataView,
): IDrillEvent {
    const dv = new DataViewFacade(dataView);
    const measureHeaderItem: IMeasureDescriptor = dv.measureDescriptor(itemContext.localIdentifier);
    if (!measureHeaderItem) {
        throw new Error("The metric uri has not been found in execution response!");
    }

    const intersectionElement: IDrillEventIntersectionElement = {
        header: measureHeaderItem,
    };
    const drillContext: IDrillEventContextHeadline = {
        type: VisualizationTypes.HEADLINE,
        element: itemContext.element,
        value: itemContext.value,
        intersection: [intersectionElement],
    };

    return {
        dataView,
        drillContext,
    };
}
