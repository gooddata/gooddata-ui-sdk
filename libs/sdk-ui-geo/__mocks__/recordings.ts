// (C) 2020 GoodData Corporation

import { DataViewFacade } from "@gooddata/sdk-ui";
import { ScenarioRecording, recordedDataView } from "@gooddata/sdk-backend-mockingbird";
import { ExamplesRecordings } from "@gooddata/live-examples-workspace";
import { getGeoData } from "../src/core/geoChart/helpers/geoChart/data";
import cloneDeep from "lodash/cloneDeep";
export function recordedDataFacade(rec: ScenarioRecording): DataViewFacade {
    return DataViewFacade.for(recordedDataView(rec));
}

function createShortcut(rec: ScenarioRecording) {
    const dv = recordedDataFacade(rec);

    return {
        dv,
        geoData: getGeoData(dv, "empty value", "null value"),
    };
}

function trimmedScenario(rec: ScenarioRecording, to: number[]) {
    const copy = cloneDeep(recordedDataView(rec));

    to.forEach((trim, dim) => {
        if (trim < 0) {
            return;
        }

        const headers = copy.headerItems[dim];

        if (headers) {
            for (let i = 0; i < headers.length; i++) {
                headers[i] = headers[i].slice(0, trim);
            }
        }

        if (dim === 0) {
            // @ts-ignore
            copy.data = copy.data.slice(0, trim) as any;
        } else if (dim === 1) {
            for (let i = 0; i < copy.data.length; i++) {
                const row = copy.data[i];

                if (Array.isArray(row)) {
                    copy.data[i] = row.slice(0, trim);
                }
            }
        }

        if (copy.count[dim]) {
            copy.count[dim] = trim;
        }

        if (copy.totalCount[dim]) {
            copy.totalCount[dim] = trim;
        }
    });

    const dv = DataViewFacade.for(copy);

    return {
        dv,
        geoData: getGeoData(dv, "empty value", "null value"),
    };
}

export const RecShortcuts = {
    LocationOnly: createShortcut(ExamplesRecordings.Scenarios.GeoPushpinChart.LocationOnly),
    LocationOnlySmall: trimmedScenario(ExamplesRecordings.Scenarios.GeoPushpinChart.LocationOnly, [10]),
    LocationAndSize: createShortcut(ExamplesRecordings.Scenarios.GeoPushpinChart.LocationAndSize),
    LocationAndColor: createShortcut(ExamplesRecordings.Scenarios.GeoPushpinChart.LocationAndColor),
    LocationAndColor_Small: trimmedScenario(
        ExamplesRecordings.Scenarios.GeoPushpinChart.LocationAndColor,
        [-1, 10],
    ),
    LocationSizeAndTooltip: createShortcut(
        ExamplesRecordings.Scenarios.GeoPushpinChart.LocationAndSizeWithTooltip,
    ),
    LocationSizeAndTooltip_Small: trimmedScenario(
        ExamplesRecordings.Scenarios.GeoPushpinChart.LocationAndSizeWithTooltip,
        [-1, 10],
    ),
    LocationSizeAndColor: createShortcut(ExamplesRecordings.Scenarios.GeoPushpinChart.LocationSizeAndColor),
    LocationSizeAndColor_Small: trimmedScenario(
        ExamplesRecordings.Scenarios.GeoPushpinChart.LocationSizeAndColor,
        [-1, 10],
    ),
    LocationSegmentAndColor: createShortcut(
        ExamplesRecordings.Scenarios.GeoPushpinChart.LocationSegmentAndColor,
    ),
    LocationSegmentAndColor_Small: trimmedScenario(
        ExamplesRecordings.Scenarios.GeoPushpinChart.LocationSegmentAndColor,
        [-1, 10],
    ),
    LocationSegmentAndSize: createShortcut(
        ExamplesRecordings.Scenarios.GeoPushpinChart.LocationSegmentAndSize,
    ),
    All: createShortcut(ExamplesRecordings.Scenarios.GeoPushpinChart.LocationSegmentSizeAndColorWithTooltip),
    AllAndSmall: createShortcut(
        ExamplesRecordings.Scenarios.GeoPushpinChart.LocationSegmentSizeAndColorWithTooltipAndFilter,
    ),
};
