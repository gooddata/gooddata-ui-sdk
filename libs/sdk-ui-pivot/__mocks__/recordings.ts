// (C) 2020 GoodData Corporation

import { DataViewFacade } from "@gooddata/sdk-ui";
import { ScenarioRecording, recordedDataView, DataViewAll } from "@gooddata/sdk-backend-mockingbird";

export function recordedDataFacade(rec: ScenarioRecording, dataViewId: string = DataViewAll): DataViewFacade {
    return DataViewFacade.for(recordedDataView(rec, dataViewId));
}
