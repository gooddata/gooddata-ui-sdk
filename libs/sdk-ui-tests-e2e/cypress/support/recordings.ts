// (C) 2022 GoodData Corporation
import { getMockServer } from "./constants";

let testInfo: any;

Cypress.on("test:before:run", (_attributes, test) => {
    testInfo = test;
});

beforeEach(() => {
    const testName = `${testInfo.parent.title} - It ${testInfo.title}`;
    let heading = "TEST:";
    if (testInfo._currentRetry !== 0) {
        heading = "  retry:";
    }
    cy.task("log", `  ${heading} ${testName}`);

    /**
     * Reset scenario state of the mock server
     */
    const mockServer = getMockServer();
    if (mockServer) {
        cy.task("resetRecordingsScenarios", mockServer);
    }

    /**
     * Detect tests that need separate recordings.
     * Separate recordings are needed just for POST and PUT requests.
     * Requests are made unique by adding extra header with test name.
     */
    const separateRecordings = testName.indexOf("(SEPARATE)") !== -1;
    if (separateRecordings) {
        const separateRecordingsKey = testName.replace(/\s/g, "-");
        cy.intercept("POST", "*", (request) => {
            if (request.url.indexOf("executeAfm") === -1) {
                request.headers["X-GDC-TEST-NAME"] = separateRecordingsKey;
            }
        });
        cy.intercept("PUT", "*", (request) => {
            request.headers["X-GDC-TEST-NAME"] = separateRecordingsKey;
        });
    }
});
