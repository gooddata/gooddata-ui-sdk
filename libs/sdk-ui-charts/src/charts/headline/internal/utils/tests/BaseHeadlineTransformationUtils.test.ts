// (C) 2023 GoodData Corporation
import { describe, expect, it } from "vitest";

import { DataViewFacade, ExplicitDrill, HeadlineElementType } from "@gooddata/sdk-ui";
import { ScenarioRecording } from "@gooddata/sdk-backend-mockingbird";
import { ReferenceRecordings } from "@gooddata/reference-workspace";

import { recordedDataFacade } from "../../../../../../__mocks__/recordings.js";
import { TEST_MULTI_MEASURE_TRANSFORMATION } from "../../tests/TestData.fixtures.js";
import { createBaseHeadlineItem, getBaseHeadlineData } from "../BaseHeadlineTransformationUtils.js";
import { getExecutionData } from "../HeadlineTransformationUtils.js";

describe("BaseHeadlineTransformationUtils", () => {
    describe("getBaseHeadlineData", () => {
        it.each(TEST_MULTI_MEASURE_TRANSFORMATION)(
            "should transform '%s' correctly",
            (_test: string, recorded: ScenarioRecording, drillableItems: ExplicitDrill[] = []) => {
                const dataFacade = recordedDataFacade(recorded);
                const data = getBaseHeadlineData(dataFacade.dataView, drillableItems);
                expect(data).toMatchSnapshot();
            },
        );
    });

    describe("createBaseHeadlineItem", () => {
        const dataFacade = recordedDataFacade(
            ReferenceRecordings.Scenarios.Headline.MultiMeasuresWithThreeMeasures,
        );
        const [firstExecutionData] = getExecutionData(DataViewFacade.for(dataFacade.dataView));

        it("Should return base headline item correctly", () => {
            expect(createBaseHeadlineItem(firstExecutionData, false, "primaryValue")).toMatchSnapshot();
        });

        it.each([true, false])("Should return base headline item with drillable is %s", (expected) => {
            const baseHeadlineItem = createBaseHeadlineItem(firstExecutionData, expected, "primaryValue");
            expect(baseHeadlineItem.data.isDrillable).toBe(expected);
        });

        it.each(["primaryValue", "secondaryValue"])(
            "Should return base headline item with elementType is %s",
            (elementType: HeadlineElementType) => {
                const baseHeadlineItem = createBaseHeadlineItem(firstExecutionData, false, elementType);
                expect(baseHeadlineItem.elementType).toBe(elementType);
            },
        );
    });
});
