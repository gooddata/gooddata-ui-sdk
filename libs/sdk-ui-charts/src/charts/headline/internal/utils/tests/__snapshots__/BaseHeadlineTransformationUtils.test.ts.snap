// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseHeadlineTransformationUtils > createBaseHeadlineItem > Should return base headline item correctly 1`] = `
{
  "baseHeadlineDataItemComponent": [Function],
  "data": {
    "format": "$#,##0.00",
    "isDrillable": false,
    "localIdentifier": "m_e519fa2a_86c3_4e32_8313_0c03062348j3",
    "title": "Won",
    "value": "38310753.45",
  },
  "elementType": "primaryValue",
}
`;

exports[`BaseHeadlineTransformationUtils > getBaseHeadlineData > should transform 'multi measure with only primary measure' correctly 1`] = `
{
  "primaryItem": {
    "baseHeadlineDataItemComponent": [Function],
    "data": {
      "format": "$#,##0.00",
      "isDrillable": false,
      "localIdentifier": "m_e519fa2a_86c3_4e32_8313_0c03062348j3",
      "title": "Won",
      "value": "38310753.45",
    },
    "elementType": "primaryValue",
  },
  "secondaryItem": null,
  "tertiaryItem": null,
}
`;

exports[`BaseHeadlineTransformationUtils > getBaseHeadlineData > should transform 'multi measure with three measure (drilling on primary)' correctly 1`] = `
{
  "primaryItem": {
    "baseHeadlineDataItemComponent": [Function],
    "data": {
      "format": "$#,##0.00",
      "isDrillable": true,
      "localIdentifier": "m_e519fa2a_86c3_4e32_8313_0c03062348j3",
      "title": "Won",
      "value": "38310753.45",
    },
    "elementType": "primaryValue",
  },
  "secondaryItem": {
    "baseHeadlineDataItemComponent": [Function],
    "data": {
      "format": "$#,##0.00",
      "isDrillable": true,
      "localIdentifier": "m_87a053b0_3947_49f3_b0c5_de53fd01f050",
      "title": "Amount",
      "value": "116625456.54",
    },
    "elementType": "secondaryValue",
  },
  "tertiaryItem": {
    "baseHeadlineDataItemComponent": [Function],
    "data": {
      "format": "#,##0.0%",
      "isDrillable": false,
      "localIdentifier": "m_b4e3e3c7_ead3_4d69_8be4_23bcfe5ff7aa",
      "title": "Probability",
      "value": "0.5975012493753123",
    },
    "elementType": "secondaryValue",
  },
}
`;

exports[`BaseHeadlineTransformationUtils > getBaseHeadlineData > should transform 'multi measure with three measure (drilling on second and third measures)' correctly 1`] = `
{
  "primaryItem": {
    "baseHeadlineDataItemComponent": [Function],
    "data": {
      "format": "$#,##0.00",
      "isDrillable": true,
      "localIdentifier": "m_e519fa2a_86c3_4e32_8313_0c03062348j3",
      "title": "Won",
      "value": "38310753.45",
    },
    "elementType": "primaryValue",
  },
  "secondaryItem": {
    "baseHeadlineDataItemComponent": [Function],
    "data": {
      "format": "$#,##0.00",
      "isDrillable": true,
      "localIdentifier": "m_87a053b0_3947_49f3_b0c5_de53fd01f050",
      "title": "Amount",
      "value": "116625456.54",
    },
    "elementType": "secondaryValue",
  },
  "tertiaryItem": {
    "baseHeadlineDataItemComponent": [Function],
    "data": {
      "format": "#,##0.0%",
      "isDrillable": true,
      "localIdentifier": "m_b4e3e3c7_ead3_4d69_8be4_23bcfe5ff7aa",
      "title": "Probability",
      "value": "0.5975012493753123",
    },
    "elementType": "secondaryValue",
  },
}
`;

exports[`BaseHeadlineTransformationUtils > getBaseHeadlineData > should transform 'multi measure with three measure (drilling)' correctly 1`] = `
{
  "primaryItem": {
    "baseHeadlineDataItemComponent": [Function],
    "data": {
      "format": "$#,##0.00",
      "isDrillable": true,
      "localIdentifier": "m_e519fa2a_86c3_4e32_8313_0c03062348j3",
      "title": "Won",
      "value": "38310753.45",
    },
    "elementType": "primaryValue",
  },
  "secondaryItem": {
    "baseHeadlineDataItemComponent": [Function],
    "data": {
      "format": "$#,##0.00",
      "isDrillable": false,
      "localIdentifier": "m_87a053b0_3947_49f3_b0c5_de53fd01f050",
      "title": "Amount",
      "value": "116625456.54",
    },
    "elementType": "secondaryValue",
  },
  "tertiaryItem": {
    "baseHeadlineDataItemComponent": [Function],
    "data": {
      "format": "#,##0.0%",
      "isDrillable": false,
      "localIdentifier": "m_b4e3e3c7_ead3_4d69_8be4_23bcfe5ff7aa",
      "title": "Probability",
      "value": "0.5975012493753123",
    },
    "elementType": "secondaryValue",
  },
}
`;

exports[`BaseHeadlineTransformationUtils > getBaseHeadlineData > should transform 'multi measure with three measures' correctly 1`] = `
{
  "primaryItem": {
    "baseHeadlineDataItemComponent": [Function],
    "data": {
      "format": "$#,##0.00",
      "isDrillable": false,
      "localIdentifier": "m_e519fa2a_86c3_4e32_8313_0c03062348j3",
      "title": "Won",
      "value": "38310753.45",
    },
    "elementType": "primaryValue",
  },
  "secondaryItem": {
    "baseHeadlineDataItemComponent": [Function],
    "data": {
      "format": "$#,##0.00",
      "isDrillable": false,
      "localIdentifier": "m_87a053b0_3947_49f3_b0c5_de53fd01f050",
      "title": "Amount",
      "value": "116625456.54",
    },
    "elementType": "secondaryValue",
  },
  "tertiaryItem": {
    "baseHeadlineDataItemComponent": [Function],
    "data": {
      "format": "#,##0.0%",
      "isDrillable": false,
      "localIdentifier": "m_b4e3e3c7_ead3_4d69_8be4_23bcfe5ff7aa",
      "title": "Probability",
      "value": "0.5975012493753123",
    },
    "elementType": "secondaryValue",
  },
}
`;

exports[`BaseHeadlineTransformationUtils > getBaseHeadlineData > should transform 'multi measure with two measures (german separator)' correctly 1`] = `
{
  "primaryItem": {
    "baseHeadlineDataItemComponent": [Function],
    "data": {
      "format": "$#,##0.00",
      "isDrillable": false,
      "localIdentifier": "m_e519fa2a_86c3_4e32_8313_0c03062348j3",
      "title": "Won",
      "value": "38310753.45",
    },
    "elementType": "primaryValue",
  },
  "secondaryItem": {
    "baseHeadlineDataItemComponent": [Function],
    "data": {
      "format": "$#,##0.00",
      "isDrillable": false,
      "localIdentifier": "m_87a053b0_3947_49f3_b0c5_de53fd01f050",
      "title": "Amount",
      "value": "116625456.54",
    },
    "elementType": "secondaryValue",
  },
  "tertiaryItem": null,
}
`;

exports[`BaseHeadlineTransformationUtils > getBaseHeadlineData > should transform 'multi measure with two measures (one derived measure)' correctly 1`] = `
{
  "primaryItem": {
    "baseHeadlineDataItemComponent": [Function],
    "data": {
      "format": "$#,##0.00",
      "isDrillable": false,
      "localIdentifier": "m_e519fa2a_86c3_4e32_8313_0c03062348j3",
      "title": "Won",
      "value": "38310753.45",
    },
    "elementType": "primaryValue",
  },
  "secondaryItem": {
    "baseHeadlineDataItemComponent": [Function],
    "data": {
      "format": "$#,##0.00",
      "isDrillable": false,
      "localIdentifier": "m_e40a8329_m_e519fa2a_86c3_4e32_8313_0c03062348j3_dt_closedate_timestamp.year",
      "title": "Won Last Year",
      "value": "38310753.45",
    },
    "elementType": "secondaryValue",
  },
  "tertiaryItem": null,
}
`;

exports[`BaseHeadlineTransformationUtils > getBaseHeadlineData > should transform 'multi measure with two measures' correctly 1`] = `
{
  "primaryItem": {
    "baseHeadlineDataItemComponent": [Function],
    "data": {
      "format": "$#,##0.00",
      "isDrillable": false,
      "localIdentifier": "m_e519fa2a_86c3_4e32_8313_0c03062348j3",
      "title": "Won",
      "value": "38310753.45",
    },
    "elementType": "primaryValue",
  },
  "secondaryItem": {
    "baseHeadlineDataItemComponent": [Function],
    "data": {
      "format": "$#,##0.00",
      "isDrillable": false,
      "localIdentifier": "m_87a053b0_3947_49f3_b0c5_de53fd01f050",
      "title": "Amount",
      "value": "116625456.54",
    },
    "elementType": "secondaryValue",
  },
  "tertiaryItem": null,
}
`;
